### PR TITLE
importinto: executor part of import from select

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: ref #49883

Problem Summary:

### What changed and how does it work?
see title
### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/local/local.go [hunk 1]
```diff
@@ -517,24 +516,6 @@ func NewBackend(
 		return nil, common.NormalizeOrWrapErr(common.ErrCreatePDClient, err)
 	}
 
-	shouldCreate := true
```
Comment (D3Hunter) [inline br/pkg/lightning/backend/local/local.go:520]: moved to engine-mgr, below `Close` part too

**Comment by D3Hunter** [comment]
ci success
https://do.pingcap.net/jenkins/blue/organizations/jenkins/pingcap%2Ftidb%2Fpull_lightning_integration_test/detail/pull_lightning_integration_test/23/pipeline

**Comment by D3Hunter** [comment]
@lance6716 @GMHDBJD

File: br/pkg/lightning/backend/local/engine_mgr.go [hunk 2]
```diff
@@ -566,3 +579,24 @@ func openDuplicateDB(storeDir string) (*pebble.DB, error) {
...
+			if !os.IsNotExist(err) {
+				return err
+			}
+		} else if info.IsDir() {
+			shouldCreate = false
+		}
+	}
+
+	if shouldCreate {
+		err := os.Mkdir(config.LocalStoreDir, 0o700)
```
Comment (lance6716) [inline br/pkg/lightning/backend/local/engine_mgr.go:596]: use MkdirAll?

**Review by lance6716 — COMMENTED** [review]
7 / 25 files viewed

File: pkg/executor/import_into.go [hunk 3]
```diff
@@ -237,12 +244,106 @@ func (e *ImportIntoExec) doImport(ctx context.Context, se sessionctx.Context, di
...
+	defer func() {
+		if err := ti.Close(); err != nil {
+			logutil.Logger(ctx).Error("close importer failed", zap.Error(err))
+		}
+	}()
+	selectedRowCh := make(chan importer.QueryRow)
+	ti.SetSelectedRowCh(selectedRowCh)
+
+	var importResult *importer.JobImportResult
+	eg, egCtx := errgroup.WithContext(ctx)
```
[the remark this answers is not in the document]
Comment (D3Hunter) [reply pkg/executor/import_into.go:297]: `eg` abbr of `error group`, we use `g`, `eg` or `group` in the codebase, what do you want to comment?

File: pkg/executor/import_into.go [hunk 4]
```diff
@@ -237,12 +244,106 @@ func (e *ImportIntoExec) doImport(ctx context.Context, se sessionctx.Context, di
...
+			}
+			if chk.NumRows() == 0 {
+				break
+			}
+			for innerChunkRow := iter.Begin(); innerChunkRow != iter.End(); innerChunkRow = iter.Next() {
+				idAllocator++
+				select {
+				case selectedRowCh <- importer.QueryRow{
+					ID:   idAllocator,
+					Data: innerChunkRow.GetDatumRow(fields),
```
[the remark this answers is not in the document]
Comment (D3Hunter) [reply pkg/executor/import_into.go:323]: the memory footprint is not that large here, let gc handle it now.

File: br/pkg/lightning/backend/local/engine_mgr.go (hunk 2 above)
Comment (D3Hunter) [reply br/pkg/lightning/backend/local/engine_mgr.go:596]: copied from `local backend` ctor

File: pkg/executor/import_into.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (lance6716) [reply pkg/executor/import_into.go:297]: we have two "error group", the other one is in line 271

[the remark this answers is not in the document]
Comment (D3Hunter) [reply pkg/executor/import_into.go:297]: oh, that's for `import from file`, we will start a routine to wait until job done, for non detached mode

`import from select` doesn't use it, but need to init it.

will add a comment later

[the remark this answers is not in the document]
Comment (D3Hunter) [reply pkg/executor/import_into.go:298]: https://github.com/pingcap/tidb/pull/50341/commits/87c73b0fb6ce5c5cf6d792fae7d8cdf7e57dcc8c

File: pkg/executor/importer/chunk_process.go [hunk 5]
```diff
@@ -100,8 +101,14 @@ func (b *deliverKVBatch) add(kvs *kv.Pairs) {
 	}
 }
 
-// chunkEncoder encode data chunk(either a data file or part of a file).
-type chunkEncoder struct {
+type chunkEncoder interface {
+	init() error
+	encodeLoop(ctx context.Context) error
+	logFields() []zap.Field
```
Comment (lance6716) [inline pkg/executor/importer/chunk_process.go:107]: seems it's only used when task ends. In order to not let developer use it when task is running, I prefer we name it `summary`, `summaryFields` or something.

File: pkg/executor/importer/chunk_process.go [hunk 6]
```diff
@@ -418,6 +423,153 @@ func (p *localSortChunkProcessor) deliverLoop(ctx context.Context) error {
...
+	encodeTotalDur time.Duration
+}
+
+var _ chunkEncoder = (*queryChunkEncoder)(nil)
+
+func (*queryChunkEncoder) init() error {
+	return nil
+}
+
+func (e *queryChunkEncoder) encodeLoop(ctx context.Context) error {
```
Comment (lance6716) [inline pkg/executor/importer/chunk_process.go:452]: maybe in future we can unify it with `fileChunkEncoder`? Most logic is shared, the difference is how to read data.

**Review by lance6716 — COMMENTED** [review]
11 / 25 files viewed

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: pkg/executor/importer/chunk_process.go (hunk 6 above)
Comment (D3Hunter) [reply pkg/executor/importer/chunk_process.go:453]: https://github.com/pingcap/tidb/pull/50341/commits/6c1c867d973b3a423134c128b9fb16dbbde0632c

**Review by lance6716 — APPROVED** [review]

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Pull request merged**